include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+    }
+
+    inline void
+    init_SLM_memory(void* __boost_mem, const sycl::nd_item<1>& __self_item) const
+    {
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
+        ::std::uint8_t __k;
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        for (__k = 0; __k < __factor - 1; __k++)
+        {
+            __d_boundaries[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+        // residual
+        if (__gSize * __k + __self_lidx < __bin_hash.__boundaries.size())
+        {
+            __d_boundaries[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+    }
+
+    template <typename _T2>
+    ::std::uint32_t inline get_bin(_T2&& __value, void* __boost_mem) const
+    {
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        return (::std::upper_bound(__d_boundaries, __d_boundaries + __bin_hash.__boundaries.size(),
+                                   ::std::forward<_T2>(__value)) -
+                __d_boundaries) -
+               1;
+    }
+
+    template <typename _T2>
+    bool inline is_valid(const _T2& __value, void* __boost_mem) const
+    {
+        __boundary_type* __d_boundaries = static_cast<__boundary_type*>(__boost_mem);
+        return (__value >= __d_boundaries[0]) && (__value < __d_boundaries[__bin_hash.__boundaries.size()]);
+    }
+};
+
+template <typename _HistAccessor, typename _OffsetT, typename _Size>
+inline void
+__clear_wglocal_histograms(const _HistAccessor& __local_histogram, const _OffsetT& __offset, const _Size& __num_bins,
+                           const sycl::nd_item<1>& __self_item)
+{
+    ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+    ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+    ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__num_bins, __gSize);
+    ::std::uint8_t __k;
+
+    for (__k = 0; __k < __factor - 1; __k++)
+    {
+        __local_histogram[__offset + __gSize * __k + __self_lidx] = 0;
+    }
+    // residual
+    if (__gSize * __k + __self_lidx < __num_bins)
+    {
+        __local_histogram[__offset + __gSize * __k + __self_lidx] = 0;
+    }
+    __dpl_sycl::__group_barrier(__self_item);
+}
+
+template <typename _BinIdxType, typename _Iter1, typename _HistReg, typename _BinFunc>
+inline void
+__accum_local_register_iter(const _Iter1& __in_acc, const ::std::size_t& __index, _HistReg* __histogram,
+                            _BinFunc __func, void* __boost_mem)
+{
+    const auto& __x = __in_acc[__index];
+    if (__func.is_valid(__x))
+    {
+        _BinIdxType c = __func.get_bin(__x, __boost_mem);
+        __histogram[c]++;
```

Comment:
In given case, I trust to the modern compiler that it makes in-place increment here.... But, several years was a rule to have pre-increment to avoid a temporary variable creation (even in such case  - a kind of style..... I don't insist).